moodle/Sniffs/PHPUnit/TestClassesFinalSniff.php

```diff
@@ -0,0 +1,126 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace MoodleHQ\MoodleCS\moodle\Sniffs\PHPUnit;
+
+// phpcs:disable moodle.NamingConventions
+
+use MoodleHQ\MoodleCS\moodle\Util\MoodleUtil;
+use PHP_CodeSniffer\Sniffs\Sniff;
+use PHP_CodeSniffer\Files\File;
+use PHPCSUtils\Utils\ObjectDeclarations;
+
+/**
+ * Checks that test classes are declared as final.
+ *
+ * @package    local_codechecker
+ * @copyright  2024 Andrew Lyons <adrew@nicols.co.uk>
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class TestClassesFinalSniff implements Sniff {
+    public function register() {
+        return [
+            T_OPEN_TAG,
+        ];
+    }
+
+    public function process(File $file, $pointer) {
+        // If the file is not a unit test file, nothing to check.
+        if (!MoodleUtil::isUnitTest($file) && !MoodleUtil::isUnitTestRunning()) {
+            return; // @codeCoverageIgnore
+        }
+
+        // If we aren't checking Moodle 4.4dev (404) and up, nothing to check.
+        // Make and exception for codechecker phpunit tests, so they are run always.
+        if (!MoodleUtil::meetsMinimumMoodleVersion($file, 404) && !MoodleUtil::isUnitTestRunning()) {
+            return; // @codeCoverageIgnore
+        }
+
+        // Get the file tokens, for ease of use.
+        $tokens = $file->getTokens();
+
+        // We only want to do this once per file.
+        $prevopentag = $file->findPrevious(T_OPEN_TAG, $pointer - 1);
+        if ($prevopentag !== false) {
+            return; // @codeCoverageIgnore
+        }
+
+        $cStart = $pointer;
+        while ($cStart = $file->findNext(T_CLASS, $cStart + 1)) {
+            if (MoodleUtil::isUnitTestCaseClass($file, $cStart) === false) {
+                // This class does not relate to a unit test.
+                continue;
+            }
+            $className = ObjectDeclarations::getName($file, $cStart);
+            if (substr($className, -5) !== '_test') {
+                continue;
+            }
+
+            $classInfo = ObjectDeclarations::getClassProperties($file, $cStart);
+
+            if ($classInfo['is_final']) {
+                // Already final.
+                continue;
+            }
+
+            if ($classInfo['is_abstract']) {
+                // See if this class has any abstract methods.
+                $mStart = $cStart + 1;
+                $hasAbstractMethod = false;
+
+                while ($mStart = $file->findNext(T_ABSTRACT, $mStart, $tokens[$cStart]['scope_closer']) !== false) {
+                    $hasAbstractMethod = true;
+                    break;
+                }
+                if ($hasAbstractMethod) {
+                    $file->addWarning(
```

Comment:
My brain is going to explode! Sure that there is a reason for the split of this and the `else` few lines below, but with the message being exactly the same, I'm missing it. Surely I need a siesta!

Reply:
If the test class is abstract, and contains abstract methods, then we cannot fix it automatically.
If the test class is abstract, but does not contains abstract methods, then we can make it final automatically (though it may well break something).

Really if they're abstract they should be converted to a testcase, but we can't automate that.

Reply:
Happy to make a change either way. It's pretty commonplace to have the same message for those that can be auto-fixed and those that cannot though from what I've seen.

Reply:
Ok, I buy your argument, though there is still something that "resonates" in my brain (but sure that it's my brain's defect, so all ok!).